### PR TITLE
UART RX implementation on PIC32 devices

### PR DESCRIPTION
**Pull request by francois-berder**
### Contribution description

This PR finally implements UART RX on PIC32 devices and makes it possible to run `examples/default` on pic32-clicker and pic32-wifire.

#### GPIO

Have a look at PR #12475. GPIOs are now working very similarly that on STM32.

#### Interrupts

Interrupts were refactored: we assume devices use EIC (that is true since Malta board was removed from RIOT) and the implementation is similar to RISCV FE310 CPU EIC.
I also found out that updating the MIPS toolchain completely broke interrupts due to a missing file (`mips_excpt_isr.S`) in 2018.09-03 that used to be present in 2016.05-03 version. So `mips_excpt_isr.S` is now present in `cpu/misp32r2_common`.

#### UART

UART peripheral driver was also refactored to make it more similar to STM32 UART driver implementation. UART configuration is now defined in board `periph_conf.h` file.
The major novelty of this PR is that is now possible to receive data over serial.

#### miscellaneous

`pm_reboot` was implemented in `cpu/mips_pic32_common` such that all commands available when running `examples/default` are now working on pic32-clicker and pic32-wifire.

### Testing procedure

I ran `examples/default` and checked that I could execute commands on pic32-clicker and pic32-wifire.

### Issues/PRs references

Depends on ~~#12475~~.

### CODE REVIEW

**Comment by francois-berder** [comment]
This PR is now ready for review.
@benpicco Could you remove the "waiting for other PR" label ?

**Comment by francois-berder** [comment]
Interestingly, a previous murdock [build](https://ci.riot-os.org/RIOT-OS/RIOT/13094/dd77b56be0983f541d1d811b9212f3dc3c448a78/output.html#error0) revealed that MIPS toolchain uses a different type for `size_t` than other toolchains.

```sh
$ echo | gcc -E -xc -include 'stddef.h' - | grep size_t
typedef long unsigned int size_t;
$ echo | arm-none-eabi-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | avr-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | mips-mti-elf-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef long unsigned int size_t;
$ echo | msp430-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | riscv-none-embed-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | xtensa-esp32-elf-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | xtensa-esp8266-elf-gcc -E -xc -include 'stddef.h' - | grep size_t
typedef unsigned int size_t;
$ echo | clang -E -xc -include 'stddef.h' - | grep size_t
typedef long unsigned int size_t;
```

```
Operating System Environment
-----------------------------
         Operating System: "Ubuntu" "18.04.4 LTS (Bionic Beaver)"
                   Kernel: Linux 5.3.0-28-generic x86_64 x86_64
             System shell: /bin/dash (probably dash)
             make's shell: /bin/dash (probably dash)

Installed compiler toolchains
-----------------------------
               native gcc: gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
        arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 8-2019-q3-update) 8.3.1 20190703 (release) [gcc-8-branch revision 273027]
                  avr-gcc: avr-gcc (GCC) 5.4.0
         mips-mti-elf-gcc: mips-mti-elf-gcc (Codescape GNU Tools 2018.09-03 for MIPS MTI Bare Metal) 6.3.0
               msp430-gcc: msp430-gcc (GCC) 4.6.3 20120301 (mspgcc LTS 20120406 unpatched)
     riscv-none-embed-gcc: riscv-none-embed-gcc (GNU MCU Eclipse RISC-V Embedded GCC, 64-bit) 8.2.0
     xtensa-esp32-elf-gcc: xtensa-esp32-elf-gcc (crosstool-NG crosstool-ng-1.22.0-80-g6c4433a5) 5.2.0
   xtensa-esp8266-elf-gcc: xtensa-esp8266-elf-gcc (crosstool-NG crosstool-ng-1.22.0-80-g6c4433a5) 5.2.0
                    clang: clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)

Installed compiler libs
-----------------------
     arm-none-eabi-newlib: "3.1.0"
      mips-mti-elf-newlib: "2.5.0"
  riscv-none-embed-newlib: "3.0.0"
  xtensa-esp32-elf-newlib: "2.2.0"
xtensa-esp8266-elf-newlib: "2.2.0"
                 avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                   ccache: ccache version 3.4.1
                    cmake: cmake version 3.10.2
                 cppcheck: Cppcheck 1.82
                  doxygen: 1.8.13
                      git: git version 2.17.1
                     make: GNU Make 4.1
                  openocd: Open On-Chip Debugger 0.10.0
                   python: Python 2.7.17
                  python2: Python 2.7.17
                  python3: Python 3.6.9
                   flake8: 3.5.0 (mccabe: 0.6.1, pycodestyle: 2.3.1, pyflakes: 1.6.0) CPython 3.6.9 on Linux
               coccinelle: spatch version 1.0.4 with Python support and with PCRE support
```

**Comment by francois-berder** [comment]
In the last Murdock [build](https://ci.riot-os.org/RIOT-OS/RIOT/13094/4cc65fea189116bb33b9fa1115eed431ddadddc1/output.html), all lwip related tests failed because it could not clone lwip repository. However, I was able to do it on my machine.

File: cpu/mips_pic32_common/Makefile.include [hunk 1]
```diff
@@ -1,3 +1,8 @@
 include $(RIOTCPU)/mips32r2_common/Makefile.include
 
+CFLAGS += -DCPU_FAM_$(call uppercase_and_underscore,$(CPU_FAM))
+
+# mips32 needs periph_timer for its gettimeofday() implementation
+USEMODULE += periph_timer
```
Comment (aabadie) [inline cpu/mips_pic32_common/Makefile.include]: This should be moved to `cpu/mips_pic32_common/Makefile.dep`

File: cpu/mips_pic32mx/Makefile.include [hunk 2]
```diff
@@ -1,3 +1,5 @@
+CPU_ARCH = m4k
```
Comment (aabadie) [inline cpu/mips_pic32mx/Makefile.include]: I would prefer CPU_ARCH/CPU_FAM set in `Makefile.features`, same for `mips_pic32mz`

File: cpu/mips32r2_common/include/eic.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * @param[in] vecNum
+ */
+void eic_clear_flag(int vecNum);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* EIC_H */
+/** @} */
```
Comment (aabadie) [inline cpu/mips32r2_common/include/eic.h]: missing newline

Comment (francois-berder) [reply cpu/mips32r2_common/include/eic.h]: Done

File: cpu/mips_pic32_common/Makefile.include (hunk 1 above)
Comment (francois-berder) [reply cpu/mips_pic32_common/Makefile.include]: Done

File: cpu/mips_pic32mx/Makefile.include (hunk 2 above)
Comment (francois-berder) [reply cpu/mips_pic32mx/Makefile.include]: I put it in these files because when I grep CPU_ARCH/CPU_FAM, they are always defined in `Makefile.include`.

**Comment by francois-berder** [comment]
> This is a major refactoring and unfortunately, I don't have the hardware to test this.

Indeed but I could split this PR in several parts:

1. EIC refactoring
2. UART refactoring
3. `pm_reboot` implementation for PIC32 devices

Let me know what is preferred.

Concerning the lack of hardware, I do have a spare WiFire board. Send me an email if you are interested.

File: cpu/mips_pic32mx/Makefile.include (hunk 2 above)
Comment (aabadie) [reply cpu/mips_pic32mx/Makefile.include]: After #13236 and #13174 are merged this will no longer be the case ;)

Comment (francois-berder) [reply cpu/mips_pic32mx/Makefile.include]: done

**Comment by aabadie** [comment]
@francois-berder, I can't test this PR and I don't who is able to test except you. If you could paste the output of the build/flash/term of `example/default`, call `help`, `reboot`, that would be enough for me to merge this PR.

**Comment by francois-berder** [comment]
In this case, here are some logs for both clicker and wifire:
[build-clicker.log](https://github.com/RIOT-OS/RIOT/files/4395541/build-clicker.log)
[build-wifire.log](https://github.com/RIOT-OS/RIOT/files/4395542/build-wifire.log)
[term-clicker.log](https://github.com/RIOT-OS/RIOT/files/4395543/term-clicker.log)
[term-wifire.log](https://github.com/RIOT-OS/RIOT/files/4395544/term-wifire.log)

**Pull request merged**